In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [3]:
# Create path
path = r'G:\My Drive\CareerFoundry\Python Projects\2023-10 Instacart Basket Analysis'

In [4]:
# Import orders_prods_merged_new_var.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new_var.pkl'))

In [5]:
ords_prods_merge.head()

order_id  user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329        1                 1                   2                  8   
1   2398795        1                 2                   3                  7   
2    473747        1                 3                   3                 12   
3   2254736        1                 4                   4                  7   
4    431534        1                 5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

In [6]:
ords_prods_merge.shape

(32404859, 18)

## 2. Find the aggregated mean of the “order_number” column grouped by “department_id” for entire df.

In [7]:
# Split the data into groups based on “department_id.”
# Apply the agg() function to each group to obtain the mean values for the “order_number” column.

ords_prods_merge.groupby('department_id').agg({'number_of_orders': ['mean']})

number_of_orders
                          mean
department_id                 
1                    15.457838
2                    17.277920
3                    17.170395
4                    17.811403
5                    15.215751
6                    16.439806
7                    17.225802
8                    15.340650
9                    15.895474
10                   20.197148
11                   16.170638
12                   15.887671
13                   16.583536
14                   16.773669
15                   16.165037
16                   17.665606
17                   15.694469
18                   19.310397
19                   17.177343
20                   16.473447
21                   22.902379

## 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

### In the subset, deparments 16 and 19 had the highest average, but when using the entire df, departments 21 and 10 sit at the top. In addition, the overall averages of each department have decreased with the inclusion of the other 31M records (save for dept. 17).

## 4. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [8]:
# Split df into groups using user_id
# Apply transform() to number_of_orders col 
# New 'max_order' is now created

In [9]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_orders'].transform(np.max)

In [10]:
ords_prods_merge.head(15)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7     9.0   both   
12          1         Soda        77              7     9.0   both   
13          1         Soda        77              7     9.0   both   
14          1         Soda        77              7     9.0   both   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy  Least busy days        Average orders   
2   Mid-range product  Regularly busy  Least busy days           Most orders   
3   Mid-range product      Least busy  Least busy days        Average orders   
4   Mid-range product      Least busy  Least busy days           Most orders   
5   Mid-range product  Regularly busy   Regularly busy        Average orders   
6   Mid-range product  Regularly 

In [11]:
# Check code execution by looking at only a limited number of necessary columns 

ords_prods_merge[['user_id', 'number_of_orders', 'max_order']].head(20)

user_id  number_of_orders  max_order
0         1                 1         10
1         1                 2         10
2         1                 3         10
3         1                 4         10
4         1                 5         10
5         1                 6         10
6         1                 7         10
7         1                 8         10
8         1                 9         10
9         1                10         10
10       15                15         22
11       15                17         22
12       15                18         22
13       15                21         22
14       15                22         22
15       19                 2          9
16       19                 5          9
17       19                 7          9
18       21                10         33
19       31                10         20

In [12]:
# Create a flag to assign loyalty label using loc()

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
# Check output with value_counts()

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [17]:
# Check code execution by looking at only a limited number of necessary columns 

ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_orders']].head(60)

user_id      loyalty_flag  number_of_orders
0         1      New customer                 1
1         1      New customer                 2
2         1      New customer                 3
3         1      New customer                 4
4         1      New customer                 5
5         1      New customer                 6
6         1      New customer                 7
7         1      New customer                 8
8         1      New customer                 9
9         1      New customer                10
10       15  Regular customer                15
11       15  Regular customer                17
12       15  Regular customer                18
13       15  Regular customer                21
14       15  Regular customer                22
15       19      New customer                 2
16       19      New customer                 5
17       19      New customer                 7
18       21  Regular customer                10
19       31  Regular customer                10
20       31  Regular customer                17
21       43  Regular customer                 6
22       43  Regular customer                 9
23       52  Regular customer                 1
24       52  Regular customer                 2
25       52  Regular customer                 3
26       52  Regular customer                 5
27       52  Regular customer                 7
28       52  Regular customer                 9
29       52  Regular customer                10
30       52  Regular customer                11
31       52  Regular customer                12
32       52  Regular customer                15
33       52  Regular customer                17
34       52  Regular customer                18
35       52  Regular customer                19
36       52  Regular customer                21
37       67  Regular customer                 1
38       67  Regular customer                 2
39       67  Regular customer                 3
40       67  Regular customer                 4
41       67  Regular customer                 5
42       67  Regular customer                 6
43       67  Regular customer                 8
44       67  Regular customer                 9
45       67  Regular customer                10
46       67  Regular customer                11
47       67  Regular customer                12
48       67  Regular customer                13
49       67  Regular customer                14
50       67  Regular customer                16
51       67  Regular customer                18
52       67  Regular customer                19
53       67  Regular customer                21
54       67  Regular customer                22
55       67  Regular customer                23
56       81      New customer                 3
57       81      New customer                 5
58       82  Regular customer                 1
59       82  Regular customer                 3

## 5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer).

In [18]:
# Perfom multiple arguments by adding more arguments (mean, min, max) to find basic statistics of loyalty category

In [19]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### NEW CUSTOMERS average the highest spend per order at 13.29, while LOYAL CUSTOMERS average the lowest spend per order at 10.39 and REGULAR CUSTOMERS average 12.50 per order.

## 6. Create a spending flag for each user based on the average price across all their orders.

In [20]:
# Find the average price per item for each user and create new column - 'avg_item_price'

ords_prods_merge['avg_item_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [21]:
ords_prods_merge.head(15)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  ...  department_id  prices  _merge  \
0           0         Soda  ...              7     9.0    both   
1           1         Soda  ...              7     9.0    both   
2           1         Soda  ...              7     9.0    both   
3           1         Soda  ...              7     9.0    both   
4           1         Soda  ...              7     9.0    both   
5           1         Soda  ...              7     9.0    both   
6           1         Soda  ...              7     9.0    both   
7           1         Soda  ...              7     9.0    both   
8           1         Soda  ...              7     9.0    both   
9           1         Soda  ...              7     9.0    both   
10          0         Soda  ...              7     9.0    both   
11          1         Soda  ...              7     9.0    both   
12          1         Soda  ...              7     9.0    both   
13          1         Soda  ...              7     9.0    both   
14          1         Soda  ...              7     9.0    both   

      price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy  Least busy days        Average orders   
2   Mid-range product  Regularly busy  Least busy days           Most orders   
3   Mid-range product      Least busy  Least busy days        Average orders   
4   Mid-range product      Least busy  Least busy days           Most orders   
5   Mid-range product  Regularly busy   Regularly busy        Average orders   
6   Mid-range product  Regularly busy     Busiest days           Most orders   
7   Mid-range pro

In [22]:
# Create a flag to assign spending label using loc()

In [23]:
ords_prods_merge.loc[ords_prods_merge['avg_item_price'] < 10, 'spending_flag'] = 'Low spender'

In [24]:
ords_prods_merge.loc[ords_prods_merge['avg_item_price'] >= 10, 'spending_flag'] = 'High spender'

In [25]:
# Check output with value_counts()

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [26]:
# Check code execution by looking at only a limited number of necessary columns 

ords_prods_merge[['user_id', 'avg_item_price', 'spending_flag']].head(60)

user_id  avg_item_price spending_flag
0         1        6.367797   Low spender
1         1        6.367797   Low spender
2         1        6.367797   Low spender
3         1        6.367797   Low spender
4         1        6.367797   Low spender
5         1        6.367797   Low spender
6         1        6.367797   Low spender
7         1        6.367797   Low spender
8         1        6.367797   Low spender
9         1        6.367797   Low spender
10       15        3.980556   Low spender
11       15        3.980556   Low spender
12       15        3.980556   Low spender
13       15        3.980556   Low spender
14       15        3.980556   Low spender
15       19        8.372059   Low spender
16       19        8.372059   Low spender
17       19        8.372059   Low spender
18       21        7.012683   Low spender
19       31        8.137793   Low spender
20       31        8.137793   Low spender
21       43        7.444898   Low spender
22       43        7.444898   Low spender
23       52        6.356287   Low spender
24       52        6.356287   Low spender
25       52        6.356287   Low spender
26       52        6.356287   Low spender
27       52        6.356287   Low spender
28       52        6.356287   Low spender
29       52        6.356287   Low spender
30       52        6.356287   Low spender
31       52        6.356287   Low spender
32       52        6.356287   Low spender
33       52        6.356287   Low spender
34       52        6.356287   Low spender
35       52        6.356287   Low spender
36       52        6.356287   Low spender
37       67        7.656790   Low spender
38       67        7.656790   Low spender
39       67        7.656790   Low spender
40       67        7.656790   Low spender
41       67        7.656790   Low spender
42       67        7.656790   Low spender
43       67        7.656790   Low spender
44       67        7.656790   Low spender
45       67        7.656790   Low spender
46       67        7.656790   Low spender
47       67        7.656790   Low spender
48       67        7.656790   Low spender
49       67        7.656790   Low spender
50       67        7.656790   Low spender
51       67        7.656790   Low spender
52       67        7.656790   Low spender
53       67        7.656790   Low spender
54       67        7.656790   Low spender
55       67        7.656790   Low spender
56       81        8.165385   Low spender
57       81        8.165385   Low spender
58       82        6.441221   Low spender
59       82        6.441221   Low spender

## 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior.

In [27]:
# Find the 'days_since_prior_order' median for each user and create new column - 'prior_order_median'

ords_prods_merge['prior_order_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [28]:
ords_prods_merge.head(15)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  ...  _merge    price_range_loc     busiest_day  \
0           0         Soda  ...    both  Mid-range product  Regularly busy   
1           1         Soda  ...    both  Mid-range product  Regularly busy   
2           1         Soda  ...    both  Mid-range product  Regularly busy   
3           1         Soda  ...    both  Mid-range product      Least busy   
4           1         Soda  ...    both  Mid-range product      Least busy   
5           1         Soda  ...    both  Mid-range product  Regularly busy   
6           1         Soda  ...    both  Mid-range product  Regularly busy   
7           1         Soda  ...    both  Mid-range product  Regularly busy   
8           1         Soda  ...    both  Mid-range product  Regularly busy   
9           1         Soda  ...    both  Mid-range product      Least busy   
10          0         Soda  ...    both  Mid-range product  Regularly busy   
11          1         Soda  ...    both  Mid-range product  Regularly busy   
12          1         Soda  ...    both  Mid-range product  Regularly busy   
13          1         Soda  ...    both  Mid-range product  Regularly busy   
14          1         Soda  ...    both  Mid-range product  Regularly busy   

       busiest_days busiest_period_of_day max_order      loyalty_flag  \
0    Regularly busy        Average orders        10      New customer   
1   Least busy days        Average orders        10      New customer   
2   Least busy days           Most orders        10      New customer   
3   Least busy days        Average orders        10      New customer   
4   Least busy days           Most orders        10      New customer   
5    Regularly busy        

In [29]:
# Create a flag to assign order_frequency label using loc()

In [30]:
ords_prods_merge.loc[ords_prods_merge['prior_order_median'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [31]:
ords_prods_merge.loc[(ords_prods_merge['prior_order_median'] > 10) & (ords_prods_merge['prior_order_median'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [32]:
ords_prods_merge.loc[ords_prods_merge['prior_order_median'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [33]:
ords_prods_merge.head(15)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  ...    price_range_loc     busiest_day  \
0           0         Soda  ...  Mid-range product  Regularly busy   
1           1         Soda  ...  Mid-range product  Regularly busy   
2           1         Soda  ...  Mid-range product  Regularly busy   
3           1         Soda  ...  Mid-range product      Least busy   
4           1         Soda  ...  Mid-range product      Least busy   
5           1         Soda  ...  Mid-range product  Regularly busy   
6           1         Soda  ...  Mid-range product  Regularly busy   
7           1         Soda  ...  Mid-range product  Regularly busy   
8           1         Soda  ...  Mid-range product  Regularly busy   
9           1         Soda  ...  Mid-range product      Least busy   
10          0         Soda  ...  Mid-range product  Regularly busy   
11          1         Soda  ...  Mid-range product  Regularly busy   
12          1         Soda  ...  Mid-range product  Regularly busy   
13          1         Soda  ...  Mid-range product  Regularly busy   
14          1         Soda  ...  Mid-range product  Regularly busy   

       busiest_days busiest_period_of_day max_order      loyalty_flag  \
0    Regularly busy        Average orders        10      New customer   
1   Least busy days        Average orders        10      New customer   
2   Least busy days           Most orders        10      New customer   
3   Least busy days        Average orders        10      New customer   
4   Least busy days           Most orders        10      New customer   
5    Regularly busy        Average orders        10      New customer   
6      Busiest days           Most orders        10      New customer   
7      Bu

In [34]:
# Check output with value_counts()

ords_prods_merge['order_freq_flag'].value_counts(dropna = False)

order_freq_flag
Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: count, dtype: int64

In [35]:
# Check code execution by looking at only a limited number of necessary columns 

ords_prods_merge[['user_id', 'prior_order_median', 'order_freq_flag']].head(60)

user_id  prior_order_median        order_freq_flag
0         1                20.0       Regular customer
1         1                20.0       Regular customer
2         1                20.0       Regular customer
3         1                20.0       Regular customer
4         1                20.0       Regular customer
5         1                20.0       Regular customer
6         1                20.0       Regular customer
7         1                20.0       Regular customer
8         1                20.0       Regular customer
9         1                20.0       Regular customer
10       15                 8.0      Frequent customer
11       15                 8.0      Frequent customer
12       15                 8.0      Frequent customer
13       15                 8.0      Frequent customer
14       15                 8.0      Frequent customer
15       19                 6.0      Frequent customer
16       19                 6.0      Frequent customer
17       19                 6.0      Frequent customer
18       21                 7.0      Frequent customer
19       31                 6.0      Frequent customer
20       31                 6.0      Frequent customer
21       43                10.0      Frequent customer
22       43                10.0      Frequent customer
23       52                 8.0      Frequent customer
24       52                 8.0      Frequent customer
25       52                 8.0      Frequent customer
26       52                 8.0      Frequent customer
27       52                 8.0      Frequent customer
28       52                 8.0      Frequent customer
29       52                 8.0      Frequent customer
30       52                 8.0      Frequent customer
31       52                 8.0      Frequent customer
32       52                 8.0      Frequent customer
33       52                 8.0      Frequent customer
34       52                 8.0      Frequent customer
35       52                 8.0      Frequent customer
36       52                 8.0      Frequent customer
37       67                 7.0      Frequent customer
38       67                 7.0      Frequent customer
39       67                 7.0      Frequent customer
40       67                 7.0      Frequent customer
41       67                 7.0      Frequent customer
42       67                 7.0      Frequent customer
43       67                 7.0      Frequent customer
44       67                 7.0      Frequent customer
45       67                 7.0      Frequent customer
46       67                 7.0      Frequent customer
47       67                 7.0      Frequent customer
48       67                 7.0      Frequent customer
49       67                 7.0      Frequent customer
50       67                 7.0      Frequent customer
51       67                 7.0      Frequent customer
52       67                 7.0      Frequent customer
53       67                 7.0      Frequent customer
54       67                 7.0      Frequent customer
55       67                 7.0      Frequent customer
56       81                30.0  Non-frequent customer
57       81                30.0  Non-frequent customer
58       82                11.0       Regular customer
59       82                11.0       Regular customer

### Column names and values have been added, and code has been executed correctly.

In [36]:
#Export orders_prods_merged to prepared data as pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_grp_agg.pkl'))

In [38]:
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [39]:
ords_prods_merge['prices'].max()

25.0